src/Components/WebAssembly/WebAssembly.Authentication/src/Interop/AuthenticationService.ts

```diff
@@ -253,21 +273,61 @@ class OidcAuthorizeService implements AuthorizeService {
+    static _initialized: Promise<void>;
     static instance: OidcAuthorizeService;
+    static _pendingOperations: { [key: string]: Promise<AuthenticationResult> | undefined } = {}
 
     public static async init(settings: UserManagerSettings & AuthorizeServiceSettings) {
         // Multiple initializations can start concurrently and we want to avoid that.
         // In order to do so, we create an initialization promise and the first call to init
         // tries to initialize the app and sets up a promise other calls can await on.
         if (!AuthenticationService._initialized) {
-            this._initialized = (async () => {
-                const userManager = await this.createUserManager(settings);
+            AuthenticationService._initialized = AuthenticationService.InitializeCore(settings);
+
+            await AuthenticationService._initialized;
+        }
+
+        return AuthenticationService._initialized;
+    }
+
+    public static handleCallback() {
+        return AuthenticationService.InitializeCore();
+    }
+
+    private static async InitializeCore(settings?: UserManagerSettings & AuthorizeServiceSettings) {
+        let finalSettings = settings || AuthenticationService.resolveCachedSettings();
+        if (!settings && finalSettings) {
+            const userManager = AuthenticationService.createUserManagerCore(finalSettings);
+
+            if (window.parent != window && !window.opener && (window.frameElement &&
+                location.href.startsWith(userManager.settings.redirect_uri!))) {
+                // If we are inside a hidden iframe, try completing the sign in early.
                 AuthenticationService.instance = new OidcAuthorizeService(userManager);
-            })();
+                AuthenticationService._initialized = (async (): Promise<void> => {
+                    await AuthenticationService.instance.completeSignIn(location.href);
+                    return;
+                })();
+            }
+
+            return;
+        } else if (settings) {
+            const userManager = await AuthenticationService.createUserManager(settings);
+            AuthenticationService.instance = new OidcAuthorizeService(userManager);
+        } else {
+            // HandleCallback gets called unconditionally, so we do nothing for normal paths.
+            // Cached settings are only used on handling the redirect_uri path and if the settings are not there
+            // the app will fallback to the default logic for handling the redirect.
         }
+    }
 
-        await this._initialized;
+    private static resolveCachedSettings(): UserManagerSettings | undefined {
+        let finalSettings: UserManagerSettings | undefined;
+        const cachedSettings = window.sessionStorage.getItem(`${AuthenticationService._infrastructureKey}.CachedAuthSettings`);
+        if (cachedSettings) {
+            finalSettings = JSON.parse(cachedSettings);
+        }
+
+        return finalSettings;
```

Comment:
This logic seems weird. It looks like it reduces to something like the following. Could you check it?

```suggestion
        const cachedSettings = window.sessionStorage.getItem(`${AuthenticationService._infrastructureKey}.CachedAuthSettings`);
        return cachedSettings ? JSON.parse(cachedSettings) : undefined;
```

Reply:
done